plugin.video.putio/main.py

```diff
@@ -0,0 +1,181 @@
+# MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+# GNU General Public License for more details.
+#
+# You should have received a copy of the GNU General Public License
+# along with this program. If not, see <http://www.gnu.org/licenses/>.
+
+import os
+import sys
+import urlparse
+import requests
+
+import xbmc
+import xbmcgui
+import xbmcplugin
+
+from resources.lib.helper import __lang__
+from resources.lib.helper import __settings__
+from resources.lib.helper import PutioApiHandler
+from resources.lib.helper import PutioAuthFailureException
+
+__url__ = sys.argv[0]  # base URL ('plugin://plugin.video.putio/')
+__handle__ = int(sys.argv[1])  # process handle, as a numeric string
+__args__ = urlparse.parse_qs(sys.argv[2].lstrip('?'))  # query string, ('?action=list&item=3')
+
+PUTIO_KODI_ENDPOINT = 'https://put.io/xbmc'
+RESOURCE_PATH = os.path.join(__settings__.getAddonInfo('path'), 'resources', 'media')
+
+
+def build_url(action, item):
+    return '{0}?action={1}&item={2}'.format(__url__, action, item)
+
+
+def get_resource_path(filename):
+    """Returns special path of the given filename."""
+    if not filename:
+        return
+    return os.path.join(RESOURCE_PATH, filename)
+
+
+def populate_dir(files):
+    """Fills a directory listing with put.io files."""
+    list_items = []
+    for item in files:
+        thumbnail = item.screenshot or get_resource_path('mid-folder.png')
+        li = xbmcgui.ListItem(label=item.name,
+                              label2=item.name,
+                              iconImage=thumbnail,
+                              thumbnailImage=thumbnail)
+
+        # url when a delete action is triggered
+        delete_ctx_url = build_url(action='delete', item=item.id)
+
+        if item.is_folder:
+            url = build_url(action='list', item=item.id)
+        else:  # video or audio, no other types are available here
+            url = build_url(action='play', item=item.id)
+
+            # resumetime and totaltime are needed for Kodi to decide the file as watched or not.
+            # FIXME: get total-time of the file and set to 'totaltime'
+            li.setProperty(key='resumetime', value=str(item.start_from))
+            li.setProperty(key='totaltime', value=str(20 * 60))
+
+            # http://kodi.wiki/view/InfoLabels
+            type_ = 'video' if item.is_video else 'audio'
+            li.setInfo(type=type_, infoLabels={'size': item.size, 'title': item.name})
+
+        li.addContextMenuItems([
+            (__lang__(32040), 'Container.Refresh'),  # refresh
+            (__lang__(32041), 'Action(ParentDir)'),  # go-up
+            (__lang__(32042), 'XBMC.RunPlugin(%s)' % delete_ctx_url),  # delete
+        ])
+        list_items.append((url, li, item.is_folder))
+
+    xbmcplugin.addDirectoryItems(handle=__handle__, items=list_items, totalItems=len(list_items))
+    xbmcplugin.addSortMethod(handle=__handle__, sortMethod=xbmcplugin.SORT_METHOD_LABEL_IGNORE_THE)
```

Comment:
please also make sure to the the correct container content type for all listings. ( see http://mirrors.kodi.tv/docs/python-docs/14.x-helix/xbmcplugin.html#-setContent )